tests/unit/test_dimension_utilities.py

```diff
@@ -364,14 +404,148 @@ def test_prefetch_dimension_variables(
+        }
+        requested_variables = {
+            '/Soil_Moisture_Retrieval_Data_AM/albedo',
+            '/Soil_Moisture_Retrieval_Data_AM/surface_flag',
+        }
+        varinfo = VarInfoFromDmr(
+            'tests/data/SC_SPL3SMP_008.dmr',
+            'SPL3SMP',
+            config_file='hoss/hoss_config.json',
+        )
+
+        self.assertEqual(
+            get_prefetch_variables(
+                url,
+                varinfo,
+                requested_variables,
+                output_dir,
+                self.logger,
+                access_token,
+                self.config,
+            ),
+            prefetch_path,
+        )
+        mock_get_opendap_nc4.assert_called_once_with(
+            url, prefetch_variables, output_dir, self.logger, access_token, self.config
+        )
+        mock_check_add_artificial_bounds.assert_called_once_with(
+            prefetch_path, prefetch_variables, varinfo, self.logger
+        )
+
+    @patch('hoss.dimension_utilities.get_coordinate_variables')
+    @patch('hoss.dimension_utilities.check_add_artificial_bounds')
+    @patch('hoss.dimension_utilities.get_opendap_nc4')
+    def test_get_prefetch_variables_with_no_anonymous_dimensions(
+        self,
+        mock_get_opendap_nc4,
+        mock_check_add_artificial_bounds,
+        mock_get_coordinate_variables,
+    ):
+        """Ensure that when a list of required variables is specified, a
+         If two coordinate variables are also not present, the opendap prefetch request will not include any
+        dimension variables.
+        """
+        prefetch_path = 'tests/data/SC_SPL3SMP_008_prefetch.nc4'
+        mock_get_opendap_nc4.return_value = prefetch_path
+        access_token = 'access'
+        output_dir = 'tests/output'
+        url = 'https://url_to_opendap_granule'
+
+        requested_variables = {
+            '/Soil_Moisture_Retrieval_Data_AM/albedo',
+            '/Soil_Moisture_Retrieval_Data_AM/surface_flag',
+        }
+        varinfo = VarInfoFromDmr(
+            'tests/data/SC_SPL3SMP_008.dmr',
+            'SPL3SMP',
+            config_file='hoss/hoss_config.json',
+        )
+        with self.subTest('No coordinate variables'):
+            mock_get_coordinate_variables.return_value = ([], [])
+            self.assertEqual(
+                get_prefetch_variables(
+                    url,
+                    varinfo,
+                    requested_variables,
+                    output_dir,
+                    self.logger,
+                    access_token,
+                    self.config,
+                ),
+                prefetch_path,
+            )
+            mock_get_coordinate_variables.assert_called_once_with(
+                varinfo,
+                requested_variables,
+            )
+            mock_get_opendap_nc4.assert_called_once_with(
+                url, set(), output_dir, self.logger, access_token, self.config
```

Comment:
Okay. Just so I am reading this properly...

If there are no normal dimensions or coordinates that are found for required variables, we make a request to OPeNDAP that has nothing specified in the DAP4 constraint expression? If so, that's not great - because that implies there's nothing for us to prefetch, and that we shouldn't be trying to do one. Making a request to OPeNDAP without a constraint expression will mean OPeNDAP returns the entire granule.

Just to help trace the code through:

* This is showing we're calling `get_opendap_nc4` with an empty set for the `required_variables` argument ([see here](https://github.com/nasa/harmony-opendap-subsetter/blob/main/hoss/utilities.py#L40)).
* Those required variables [get piped into get_constraint_expression](https://github.com/nasa/harmony-opendap-subsetter/blob/main/hoss/utilities.py#L53), which [for an empty set will return an empty string](https://github.com/nasa/harmony-opendap-subsetter/blob/main/hoss/utilities.py#L81).
* Because the constraint expression is an empty string, [the request to OPeNDAP doesn't include any request data](https://github.com/nasa/harmony-opendap-subsetter/blob/main/hoss/utilities.py#L56-L59).

(This is another reason why unit tests are great - they uncovers behaviour like this!)

Reply:
The prefetch should only be called if a request is made that requires HOSS to determine index ranges (for spatial, temporal or dimension ranges). But this test is saying: sure you want us to do index-based subsetting, but we can't find anything that will allow us to calculate those index ranges.

To me, that feels like it should be an error.

Reply:
@sudha-murthy and I discussed this on a quick call. Technically, there was a small chance this same thing could have occurred prior to Sudha's changes. That said, all collections handled to-date had good dimension information, so we never got to a point where HOSS couldn't find dimensions to request in the prefetch.

Now, we have code that says: please get me dimensions or coordinates, which opens us up to a more likely case that neither dimensions nor coordinates can be found. We don't want to request the full granule back from OPeNDAP, so the proposed solution is to do a final check before making a prefetch request to OPeNDAP to ensure the `prefetch_variables` set has any elements in it. If not, an exception can be raised. (This feels like the right approach, but if it causes issues we can revert as needed)

Reply:
added exception when there are no prefetch variables
https://github.com/nasa/harmony-opendap-subsetter/commit/30e96f502411c832ee0725585bdae267f3bbd57a

Reply:
Looks good to me!